### PR TITLE
8327176: UnreferencedExecutor.java can fail on libgraal with -Xcomp

### PR DESCRIPTION
**Pull request by dougxc**
The `java/util/concurrent/Executors/UnreferencedExecutor.java` test can fail when run on libgraal and `-Xcomp` is specified. The problem is that libgraal in `-Xcomp` temporarily causes some extra memory pressure (probably due to [JDK-8310218](https://bugs.openjdk.org/browse/JDK-8310218)) which can cause recoverable OOMEs to occur (memory is recovered when the relevant libgraal compilations complete). It also seems related to async threads used for cleaning weak references when using G1 or ZGC as I cannot reproduce the failure under `-XX:+UseSerialGC`.

Installing a global `Thread.UncaughtExceptionHandler` that ignores `OutOfMemoryError`s resolves the problem.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8327176](https://bugs.openjdk.org/browse/JDK-8327176): UnreferencedExecutor.java can fail on libgraal with -Xcomp (**Bug** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18098/head:pull/18098` \
`$ git checkout pull/18098`

Update a local copy of the PR: \
`$ git checkout pull/18098` \
`$ git pull https://git.openjdk.org/jdk.git pull/18098/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18098`

View PR using the GUI difftool: \
`$ git pr show -t 18098`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18098.diff">https://git.openjdk.org/jdk/pull/18098.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18098#issuecomment-2006907789)

### CODE REVIEW

**Comment by dougxc** [comment]
I confirmed that `java/util/concurrent/Executors/UnreferencedExecutor.java` still fails if the [fix applied to ThreadContainers.java](https://github.com/openjdk/jdk/commit/ada416e66cbff6c8e631bf352acc0744c248740b#diff-1e347b9a95cc4fe81a01ca70e4122a73a65d99a69668c7567abb2a6067f8cc6dL68) is reverted.

File: test/jdk/java/util/concurrent/Executors/UnreferencedExecutor.java [hunk 1]
```diff
@@ -36,10 +36,33 @@ public class UnreferencedExecutor {
...
+                // Ignore. An unrecoverable memory leak will still
+                // cause the VM to exit.
+            } else if (e instanceof Error err) {
+                throw err;
+            } else if (e instanceof RuntimeException re) {
+                throw re;
+            } else {
+                throw new RuntimeException(e);
+            }
+        }
```
Comment (dholmes-ora) [inline test/jdk/java/util/concurrent/Executors/UnreferencedExecutor.java:58]: I'm not clear exactly what you are trying to achieve with this - is it just to not print the stacktrace when OOME occurs?

A UEH should not rethrow exceptions as they are effectively ignored as we are already at the final level of handling exceptions (the VM will print a one line warning if a UEH itself throws). If I understand what you want then I think the following would be more correct:
```
// Ignore OOME but do default handling for any other uncaught exception.
public void uncaughtException(Thread t, Throwable e) {
    if ( ! e instanceof OutOfMemoryError) {
        System.err.print("Exception in thread \"" + t.getName() + "\" ");
        e.printStackTrace(System.err);
    }
}
```

This doesn't stop the thread from terminating of course - the OOME already caused it to unwind its stack if we get this far.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

**Comment by AlanBateman** [comment]
@dougxc I added this test when fixing JDK-8308235. The point of the test is that OOME is not thrown so adding a default UEH to ignore OOME is changing the test. I wonder if this is a test that just needed to be skipped with libgraal.

**Comment by dougxc** [comment]
> @dougxc I added this test when fixing JDK-8308235. The point of the test is that OOME is not thrown so adding a default UEH to ignore OOME is changing the test. I wonder if this is a test that just needed to be skipped with libgraal.

I understood the point of that test to be that an OOME does not cause the VM to exit. As i mentioned (https://github.com/openjdk/jdk/pull/18098#issuecomment-1975231258), the test still detects the memory leak plugged by JDK-8308235.

That said, I have tried to reproduce the failure on a more recent libgraal and JDK build and am having no luck. I'll close this PR and come up with alternative solutions if the failure starts reappearing.

**Pull request closed without merging**